### DISCUSSION TITLE
netmiko.exceptions.ReadTimeout (When list of device file to big):

### DISCUSSION THREAD

**Discussion by exlr8in** [ktbyers/netmiko#2908] [Q&A]
Yes, there is quite a bit of code missing (500 lines in grand total - my testing results in success if small device list, error out in long device list)
```
import ipaddress

# subnet = "192.168.1.0/24"

# Initializing an IPv4 Network.
network = ipaddress.IPv4Network("192.168.1.0/24")

for ip in network:
	print(ip)
	print(network.network_address)
	if ip == network.network_address:
		print(type(network.network_address))
		print("i am network ID")
	if ip == network.broadcast_address:
		print("i am broadcast IP")
	get_cisco_info(clean_ip)
```

```
def get_cisco_info(clean_ip_):
    global a_connection
    global command_ip
    global search_ip
    global search_ip_re
    global found_ip
    global command_mac
    global search_mac
    global search_mac_re
    global found_mac
    global command_hostname
    global search_hostname
    global search_hostname_re
    global found_hostname

    try:
        connect_to_device(clean_ip_)

        command_run = 'show running-config'
        command_result_run = a_connection.send_command(command_run, expect_string=r'#', read_timeout=15)
        # interface_type(command_result_run)

        command_mac = 'show version | include Base Ethernet MAC Address'
        search_mac = a_connection.send_command(command_mac, expect_string=r'#', read_timeout=15)

        search_mac_re = re.search(regex_mac, search_mac)
        found_mac = search_mac_re.group()

        command_hostname = 'show running-config | include hostname'
        search_hostname = a_connection.send_command(command_hostname, expect_string=r'#', read_timeout=15)

        search_hostname_re = re.search(regex_hostname, search_hostname)
        found_hostname = search_hostname_re.group()

        a_connection.disconnect()
    except (NetmikoTimeoutException, NetmikoAuthenticationException):
        pass
```

Interactive SSH session established
Traceback (most recent call last):
  File "f.py", line 427, in <module>
    main()
  File "f.py", line 381, in main
    device_access(search_fqdn)
  File "f.py", line 273, in device_access
    get_cisco_info(clean_ip)
  File "f.py", line 185, in get_cisco_info
    connect_to_device(clean_ip_)
  File "f.py", line 165, in connect_to_device
    a_connection = ConnectHandler(**ssh_info)
  File "/home/blah/.local/lib/python3.7/site-packages/netmiko/ssh_dispatcher.py", line 365, in ConnectHandler
    return ConnectionClass(*args, **kwargs)
  File "/home/blah/.local/lib/python3.7/site-packages/netmiko/base_connection.py", line 439, in __init__
    self._open()
  File "/home/blah/.local/lib/python3.7/site-packages/netmiko/base_connection.py", line 445, in _open
    self._try_session_preparation()
  File "/home/blah/.local/lib/python3.7/site-packages/netmiko/base_connection.py", line 904, in _try_session_preparation
    self.session_preparation()
  File "/home/blah/.local/lib/python3.7/site-packages/netmiko/cisco/cisco_ios.py", line 19, in session_preparation
    self.set_terminal_width(command=cmd, pattern=cmd)
  File "/home/blah/.local/lib/python3.7/site-packages/netmiko/base_connection.py", line 1260, in set_terminal_width
    output = self.read_until_pattern(pattern=pattern)
  File "/home/blah/.local/lib/python3.7/site-packages/netmiko/base_connection.py", line 672, in read_until_pattern
    raise ReadTimeout(msg)
netmiko.exceptions.ReadTimeout: 

Pattern not detected: 'terminal width 511' in output.

Things you might try to fix this:
1. Adjust the regex pattern to better identify the terminating string. Note, in
many situations the pattern is automatically based on the network device's prompt.
2. Increase the read_timeout to a larger value.

You can also look at the Netmiko session_log or debug log for more information.

(22-08-20 14:12:13) <1> [~/a]

**Comment by ktbyers** [comment]
@exlr8in Are you using AAA for authentication and/or authorization? 

The above failure is Netmiko sent `terminal width 511` and expected it to be echoed to the screen and it wasn't echoed.

**Comment by exlr8in** [comment]
Yes, AAA (TACACS)

**Reply by ktbyers** [reply]
Are you using AAA and command authorization or just using it for authentication?

**Reply by exlr8in** [reply]
hmm yeah, I do not know the answer. I'll get back to you

**Reply by exlr8in** [reply]
btw: what if what if I am using it for either what would be a possible solution scenario fix?

**Comment by exlr8in** [comment]
Yes both for Authentication and Command Authorization.

**Reply by ktbyers** [reply]
Okay, my next question would be does the problem go away if 'command authorization' is disabled.

i.e. the next step would be test if your AAA server not responding in a reasonable amount of time is the problem.

Also are you using threads/multiprocessing or is it just a standard for-loop?

**Reply by exlr8in** [reply]
Yeah! that is going to be challenging, I do not have control of TACACS nor would it be possible to ask for disable (even temporarily) :(
No, i am not using threads/multiprocessing, it is a simple for-loop.

I guess we are stuck :(

**Reply by ktbyers** [reply]
You can probably add a big time.sleep into your for loop between devices. Maybe start with time.sleep(20) and see if it works. You can then adjust that upwards or downwards (i.e. if it is reliably working, then make the sleep shorter).

You can also increase Netmiko's global_delay_factor to 2 or 4 and see if that works.

**Reply by exlr8in** [reply]
Copy

**Reply by exlr8in** [reply]
btw: logging.getlogger("netmiko") gives me this
```
DEBUG:paramiko.transport:userauth is OK
INFO:paramiko.transport:Authentication (password) successful!
DEBUG:paramiko.transport:[chan 0] Max packet in: 32768 bytes
DEBUG:paramiko.transport:[chan 0] Max packet out: 4096 bytes
DEBUG:paramiko.transport:Secsh channel 0 opened.
DEBUG:paramiko.transport:[chan 0] Sesch channel 0 request ok
DEBUG:paramiko.transport:[chan 0] Sesch channel 0 request ok
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:write_channel: b'terminal width 511\n'
```
```
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: % Authorization failed.
```

I am trying sleep idea, will reply later - thanks

**Reply by exlr8in** [reply]
Tried both independently, Sleep and global_delay_factor and I still get the same error randomly while process the code on devices. Sometimes the failed devices pass and news appear. on a subnet of \23 i was able to get to almost the end but it is random as sometime i only get quarter into that same subnet.

Could it simply be a problem on my end and our TACACTS delay, hmmm?

**Reply by ktbyers** [reply]
Yes, it is definitely sounding like probably TACACS+ issues from what you have described (i.e. you TACACS+ server is not able to handle the load likely with the command authorization).

**Comment by exlr8in** [comment]
Nope No multithreading/multiprocessing (simple for loop) and the process elimination ask is something I don’t have access to :(
